### PR TITLE
made piezo.play accept a string-encoded song

### PR DESCRIPTION
**Pull request by danyshaanan**
@rwaldron, @BrianGenisio

Following https://github.com/julianduque/j5-songs/pull/7

Please say if additional modifications are required.

### CODE REVIEW

**Comment by danyshaanan** [comment]
I didn't manage to edit the docs without breaking the build, but this is what I've had in mind:

 https://github.com/danyshaanan/johnny-five/commit/9f2c294a68b6f633a584998079854c2bff4061aa

**Comment by BrianGenisio** [comment]
@danyshaanan You need to update the example in the `eg` folder, then run `grunt examples`.  That will produce a new docs file.

File: lib/piezo.js [hunk 1]
```diff
@@ -253,6 +253,25 @@ Piezo.prototype.play = function(tune, callback) {
       song: tune
     };
   }
+
+  if (typeof tune.song === "string") {
```
Comment (BrianGenisio) [inline lib/piezo.js]: Can this be made into a new class method? : `Piezo.stringToSong()`?

**Comment by BrianGenisio** [comment]
Looks good to me.  You'll also want to update the Wiki after it lands.

File: lib/piezo.js (hunk 1 above)
Comment (danyshaanan) [inline lib/piezo.js]: Yeah, That would also enable better testing. I'll do that.

**Comment by danyshaanan** [comment]
@BrianGenisio
Interesting flow, the `grunt examples` + `git commit` + `grunt test-examples`...
Created Piezo.stringToSong() and updated the examples.

**Comment by BrianGenisio** [comment]
LGTM :ship:

File: lib/piezo.js [hunk 2]
```diff
@@ -247,12 +247,34 @@ Piezo.prototype.frequency = function(frequency, duration) {
...
+    }
+    var lastNote = song[song.length - 1];
+    if (lastNote && lastNote[0] === note) {
+      lastNote[1] += beats;
+    } else {
+      song.push([note, beats]);
+    }
+  }
+  return song;
+};
```
Comment (rwaldron) [inline lib/piezo.js]: There is nothing instance specific about this functionality, so the operation doesn't need to be part of the public API. `Piezo.prototype.stringToSong` => `toSong` (in module scope)

File: lib/piezo.js [hunk 3]
```diff
@@ -247,12 +247,34 @@ Piezo.prototype.frequency = function(frequency, duration) {
...
+
 Piezo.prototype.play = function(tune, callback) {
   if (typeof tune !== "object") {
     tune = {
       song: tune
     };
   }
+  if (typeof tune.song === "string") {
+    tune.song = this.stringToSong(tune.song, tune.beats);
```
Comment (rwaldron) [inline lib/piezo.js]: `this.stringToSong(tune.song, tune.beats);` => `toSong(tune.song, tune.beats);`

File: test/piezo.js [hunk 4]
```diff
@@ -370,6 +370,38 @@ exports["Piezo"] = {
...
+      test.done();
+    });
+  },
+
+  stringToSong: function(test) {
+    test.expect(1);
+    var song = this.piezo.stringToSong("c d d - 672 e4 -", 0.125);
+    test.deepEqual(song, [["c",0.125],["d",0.25],["-",0.125],[672,0.125],["e4",0.125],["-",0.125]]);
+    test.done();
+  },
```
Comment (rwaldron) [inline test/piezo.js]: This doesn't need to be tested directly, only that passing a string works correctly (as tested above)

**Comment by boneskull** [comment]
:+1: for this, the current interface is a bit awkward.

**Comment by rwaldron** [comment]
@lyzadanger @julianduque @Resseguie 

Please take a look at this and provide feedback, thanks!

**Comment by Resseguie** [comment]
I like the general idea. I'll try to dig in tonight or tomorrow when I've got some hardware to test and see if I have any further suggestions.

**Comment by lyzadanger** [comment]
I'll take a peek! Glad to have a reason to get back into the swing of things.

File: lib/piezo.js (hunk 2 above)
Comment (lyzadanger) [inline lib/piezo.js]: Maybe hoist up those `var` declarations (`note`, `lastNote`)  to the top of the function for clarity as we hurtle toward the future of ES6 block-level scope?

**Comment by lyzadanger** [comment]
Just a nitpicky and possibly throwaway convention comment but otherwise LGTM :+1:

File: lib/piezo.js [hunk 5]
```diff
@@ -247,12 +247,34 @@ Piezo.prototype.frequency = function(frequency, duration) {
...
 };
 
+Piezo.prototype.stringToSong = function(stringSong, beats) {
+  beats = beats || 1;
+  var notes = stringSong.match(/\S+/g);
+  var song = [];
+  while (notes.length) {
+    var note = notes.shift();
+    if (/^[0-9]+$/.test(note)) {
+      note = parseInt(note);
```
Comment (lyzadanger) [inline lib/piezo.js]: Second (`radix`) arg for `parseInt` for safety-net's sake?

**Comment by julianduque** [comment]
LGTM  :+1: 

I would need to update https://github.com/julianduque/j5-songs/commit/220f101a66bc243d58b06517158f59665d48f628 to keep the string format after this lands :)

**Comment by danyshaanan** [comment]
All recent feedback taken into account in https://github.com/rwaldron/johnny-five/commit/1f599dfb81e799a3f0c8ea4d4db448a94161fb7d

@julianduque I was thinking of doing that, but since j5-songs and johnny-five are required side-by-side, I can't see how you can guarantee that you'll get a compatible johnny-five version. You can, however, duplicate (*shudder*) `stringToSong` and transform the song in j5-songs. Lets pick this up in j5-songs after this lands.

**Comment by julianduque** [comment]
@danyshaanan awesome, let's talk there, also added you as collaborator to the project so feel free to mess around :)

**Pull request merged**